opm/simulators/linalg/bda/BdaBridge.cpp

```diff
@@ -0,0 +1,249 @@
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h>
+#include <memory>
+
+#include <opm/simulators/linalg/bda/BdaBridge.hpp>
+#include <opm/simulators/linalg/bda/BdaResult.hpp>
+
+#define PRINT_TIMERS_BRIDGE_BRIDGE 0
+
+typedef Dune::InverseOperatorResult InverseOperatorResult;
+
+namespace Opm
+{
+
+BdaBridge::BdaBridge(bool use_gpu_, int linear_solver_verbosity, int maxit, double tolerance) : use_gpu(use_gpu_) {
+#if HAVE_CUDA
+    if(use_gpu){
+    	backend = new cusparseSolverBackend(linear_solver_verbosity, maxit, tolerance);
+    }
+#endif
+}
+
+BdaBridge::~BdaBridge(){
+#if HAVE_CUDA
+	if(use_gpu){
+		delete backend;
+	}
+#endif
+}
+
+#if HAVE_CUDA
+template <class BridgeMatrix>
+int checkZeroDiagonal(BridgeMatrix& mat) {
+	static std::vector<int> diag_indices;   // contains offsets of the diagonal nnzs
+	int numZeros = 0;
+	const int dim = 3;
+	const double zero_replace = 1e-15;
+	double *nnzs = &(mat[0][0][0][0]);
+	if(diag_indices.size() == 0){
+		int N = mat.N()*dim;
+		diag_indices.reserve(N);
+		for(typename BridgeMatrix::const_iterator r = mat.begin(); r != mat.end(); ++r){
+			for(auto c = r->begin(); c != r->end(); ++c){
+				if(r.index() == c.index()){
+					for(int rr = 0; rr < dim; ++rr){
+						// pointer arithmetic
+						int offset = (int)((long unsigned)&(mat[r.index()][c.index()][rr][rr]) - (long unsigned)nnzs); // in bytes
+						offset /= sizeof(double);  // convert offset to doubles
+						diag_indices.emplace_back(offset);
+						double val = nnzs[offset];
+						if(val == 0.0){ // could be replaced by '< 1e-30' or similar
+							nnzs[offset] = zero_replace;
+							++numZeros;
+						}
+					}
+					break;
+				}
+			}
+		}
+	}else{
+		for(int offset : diag_indices){
+			if(nnzs[offset] == 0.0){ // could be replaced by '< 1e-30' or similar
+				nnzs[offset] = zero_replace;
+				++numZeros;
+			}
+		}
+	}
+	return numZeros;
+}
+
+
+// convert matrix to blocked csr (bsr) arrays
+// if only_vals, do not convert rowPointers and colIndices
+// sparsity pattern should stay the same due to matrix-add-well-contributions
+template <class BridgeMatrix>
+void convertMatrixBsr(BridgeMatrix& mat, std::vector<double> &h_vals, std::vector<int> &h_rows, std::vector<int> &h_cols, int dim, bool only_vals) {
+	int sum_nnzs = 0;
+	int nnz = mat.nonzeroes()*dim*dim;
+
+	// copy nonzeros
+	memcpy(h_vals.data(), &(mat[0][0][0][0]), sizeof(double)*nnz);
```

Comment:
I think this currently works because we one of the ways to set up BCRSMatrix such that the data is in a contiguous array.
I do not think that this will be changed, but a comment about it should not hurt.

If the memory is continues, the shouldn't we skip the copying and just pass `reinterpret_cast<double>(&(mat[0][0][0][0])`
to all the cuda calls that copy memory? Or is there a reason that I do not see?

Reply:
No reason, I think this extra copy could be skipped, like done in `get_result`. I'll try it out.